src/fields/field_fe.cc

```diff
@@ -671,50 +668,69 @@ void FieldFE<spacedim, Value>::interpolate_intersection(ElementDataCache<double>
 
 
 template <int spacedim, class Value>
-void FieldFE<spacedim, Value>::calculate_native_values(ElementDataCache<double>::CacheData data_cache)
+void FieldFE<spacedim, Value>::calculate_element_values(ElementDataCache<double>::CacheData data_cache, bool native)
 {
-	// Same algorithm as in output of Node_data. Possibly code reuse.
-	unsigned int dof_size, data_vec_i;
-	std::vector<unsigned int> count_vector(data_vec_.size(), 0);
-	data_vec_.zero_entries();
-	std::vector<LongIdx> global_dof_indices(dh_->max_elem_dofs());
-	std::vector<LongIdx> &source_target_vec = (dynamic_cast<BCMesh*>(dh_->mesh()) != nullptr) ? source_target_mesh_elm_map_->boundary : source_target_mesh_elm_map_->bulk;
+    // Same algorithm as in output of Node_data. Possibly code reuse.
+    unsigned int data_vec_i;
+    std::vector<unsigned int> count_vector(data_vec_.size(), 0);
+    data_vec_.zero_entries();
+    std::vector<LongIdx> &source_target_vec = (dynamic_cast<BCMesh*>(dh_->mesh()) != nullptr) ? source_target_mesh_elm_map_->boundary : source_target_mesh_elm_map_->bulk;
+
+    unsigned int shift;
+    if (this->boundary_domain_) {
+        if (this->interpolation_==DataInterpolation::identic_msh) shift = this->comp_mesh_->n_elements();
+	    else shift = ReaderCache::get_mesh(reader_file_)->n_elements();
+    } else {
+        shift = 0;
+    }
 
-	// iterate through cells, assembly MPIVector
-	for (auto cell : dh_->own_range()) {
-		dof_size = cell.get_dof_indices(global_dof_indices);
-		LocDofVec loc_dofs = cell.get_loc_dof_indices();
-		data_vec_i = source_target_vec[cell.elm_idx()] * dof_size;
-		ASSERT_EQ(dof_size, loc_dofs.n_elem);
-		for (unsigned int i=0; i<dof_size; ++i, ++data_vec_i) {
-		    data_vec_.add( loc_dofs[i], (*data_cache)[ data_vec_i ] );
-		    ++count_vector[ loc_dofs[i] ];
-		}
-	}
+    for (auto cell : dh_->own_range()) {
+        LocDofVec loc_dofs = cell.get_loc_dof_indices();
+        //DebugOut() << cell.elm_idx() << " < " << source_target_vec.size() << "\n";
+        int source_idx = source_target_vec[cell.elm_idx()];
+
+        if (source_idx == (int)(Mesh::undef_idx)) { // undefined value in input data mesh
+            if ( std::isnan(default_value_) )
+                THROW( ExcUndefElementValue() << EI_Field(field_name_) );
+            for (unsigned int i=0; i<loc_dofs.n_elem; ++i) {
+                ASSERT_LT(loc_dofs[i], (LongIdx)data_vec_.size());
+                data_vec_.add( loc_dofs[i], default_value_ * this->unit_conversion_coefficient_ );
+                ++count_vector[ loc_dofs[i] ];
+            }
+        } else {
+            if (native) data_vec_i = (source_idx + shift) * loc_dofs.n_elem;
+            else data_vec_i = (source_idx + shift) * dh_->max_elem_dofs();
```

Comment:
Seems there is a bug in original code, that there should be dh_->max_elem_dofs() for the native case as well.